misc/peer-store/src/store.rs

```diff
@@ -0,0 +1,59 @@
+use libp2p_core::{Multiaddr, PeerId};
+use libp2p_swarm::FromSwarm;
+
+/// A store that
+/// - contains all observed addresses of peers;
+pub trait Store<'a> {
+    type AddressRecord;
+
+    /// Update an address record.  
+    /// Returns `true` when the address is new.  
+    fn update_address(
+        &mut self,
+        peer: &PeerId,
+        address: &Multiaddr,
+        source: AddressSource,
+        should_expire: bool,
+    ) -> bool;
+
+    /// Update an address record.  
+    /// Returns `true` when the address is new.  
+    fn update_certified_address(
+        &mut self,
+        peer: &PeerId,
+        record: libp2p_core::PeerRecord,
+        source: AddressSource,
+        should_expire: bool,
+    ) -> bool;
+
+    /// Remove an address record.
+    /// Returns `true` when the address is removed.
+    fn remove_address(&mut self, peer: &PeerId, address: &Multiaddr) -> bool;
+
+    /// How this store handles events from the swarm.
+    fn on_swarm_event(&mut self, event: &FromSwarm) -> Option<Event>;
+
+    /// Get all stored addresses of the peer.
+    fn addresses_of_peer(&self, peer: &PeerId) -> Option<impl Iterator<Item = &Multiaddr>>;
+
+    /// Get addresses of the peer that have been signed.
+    fn certified_addresses_of_peer(&self, peer: &PeerId) -> Option<impl Iterator<Item = &Multiaddr>>;
+
+    /// Trigger grabage collection for records.
+    fn check_ttl(&mut self);

```

Comment:
I think we should instead have a more generic `poll` function analogous to `NetworkBehavior::poll`, where a `Store` implementation can then implement garbage collection, cleanups, and other routine logic. Wdyt?